### PR TITLE
Added OVH DynHost support

### PR DESCRIPTION
**Pull request by xlammertink**
Hi everyone,

I added functionality to update DNS records hosted at OVH using their DynHost feature.
It would be great if you could merge this request.

Tested it with my own domain and seems to be working fine.
I'm not a perl programmer, so if you need stuff to be changed or could be better: let me know.

### CODE REVIEW

File: ddclient [hunk 1]
```diff
@@ -5264,6 +5279,84 @@ sub nic_dnsmadeeasy_update {
...
+        if ($returned =~ /good/ || $returned =~ /nochg/) {
+            $config{$h}{'ip'}     = $ip;
+            $config{$h}{'mtime'}  = $now;
+            $config{$h}{'status'} = 'good';
+            if ($returned =~ /good/) {
+                success("updating %s: good: IP address set to %s", $h, $ip);
+		    } else {
+                success("updating %s: skipped: IP address was already set to %s.", $h, $ip);
+		    }
+        } else {
```
Comment (SuperSandro2000) [inline ddclient:5352]:
```suggestion
        $url .= "$config{$h}{'server'}$config{$h}{'script'}";
        $url .= "?system=dyndns";
        $url .= "&hostname=";
        $url .= $h;
        $url .= "&myip=";
        $url .= $ip if $ip;

        my $reply = geturl(opt('proxy'), $url, "$config{$h}{'login'}", "$config{$h}{'password'}");

        if (!defined($reply) || !$reply) {
            failed("updating %s: Could not connect to %s.", $h, $config{$h}{'server'});
            next;
        }

        my @reply = split /\n/, $reply;
        my $returned = pop(@reply);
        if ($returned =~ /good/ || $returned =~ /nochg/) {
            $config{$h}{'ip'}     = $ip;
            $config{$h}{'mtime'}  = $now;
            $config{$h}{'status'} = 'good';
            if ($returned =~ /good/) {
                success("updating %s: good: IP address set to %s", $h, $ip);
            } else {
                success("updating %s: skipped: IP address was already set to %s.", $h, $ip);
		    }
        } else {
```

File: sample-etc_ddclient.conf [hunk 2]
```diff
@@ -279,9 +279,18 @@ ssl=yes					# use ssl-support.  Works with
...
 # login=your-account-email-address
 # password=your-generated-password
 # your-numeric-record-id-1,your-numeric-record-id-2,...
+
+##
+## OVH DynHost (https://ovh.com)
+##
+# protocol=ovh,
+# login=example.com-dynhostuser,
+# password=s3cuRePaSsw0rd
```
Comment (SuperSandro2000) [inline sample-etc_ddclient.conf:295]:
```suggestion
# password=your_password
```

File: ddclient [hunk 3]
```diff
@@ -5264,6 +5279,84 @@ sub nic_dnsmadeeasy_update {
...
+        my $returned = pop(@reply);
+        if ($returned =~ /good/ || $returned =~ /nochg/) {
+            $config{$h}{'ip'}     = $ip;
+            $config{$h}{'mtime'}  = $now;
+            $config{$h}{'status'} = 'good';
+            if ($returned =~ /good/) {
+                success("updating %s: good: IP address set to %s", $h, $ip);
+            } else {
+                success("updating %s: skipped: IP address was already set to %s.", $h, $ip);
+		    }
```
Comment (SuperSandro2000) [inline ddclient:5351]:
```suggestion
            }
```

File: ddclient [hunk 4]
```diff
@@ -5264,6 +5279,84 @@ sub nic_dnsmadeeasy_update {
...
+  login=dnsdomain-userid       ## The username/id registered with the service
+  password=userid-password     ## The password related to the username/id
+  fully.qualified.host         ## the hostiname registered with the service.
+
+Example ${program}.conf file entries:
+  ## single host update
+  protocol=ovh,                                   \\
+  login=example.com-dynhostuser,                  \\
+  password=s3cuRePaSsw0rd                         \\
+  test.example.com
```
Comment (SuperSandro2000) [inline ddclient:5305]:
```suggestion
  protocol=ovh,                   \\
  login=example.com-dynhostuser,  \\
  password=your_password,         \\
  test.example.com
```

File: ddclient [hunk 5]
```diff
@@ -493,6 +493,12 @@ my %variables = (
     'dondominio-common-defaults'       => {
         'server'	      => setv(T_FQDNP,  1, 0, 1, 'dondns.dondominio.com', undef),
     },
+	'ovh-common-defaults' => {
+        'server'              => setv(T_FQDNP,  1, 0, 1, 'www.ovh.com', undef),
+        'script'              => setv(T_STRING, 1, 1, 1, '/nic/update', undef),
+        'login'               => setv(T_LOGIN,  1, 0, 1, '',            undef),
+        'password'            => setv(T_PASSWD, 1, 0, 1, '',            undef),
+	},
```
Comment (SuperSandro2000) [inline ddclient:501]:
```suggestion
    'ovh-common-defaults' => {
        'server'              => setv(T_FQDNP,  1, 0, 1, 'www.ovh.com', undef),
        'script'              => setv(T_STRING, 1, 1, 1, '/nic/update', undef),
        'login'               => setv(T_LOGIN,  1, 0, 1, '',            undef),
        'password'            => setv(T_PASSWD, 1, 0, 1, '',            undef),
    },
```

File: RELEASENOTE [hunk 6]
```diff
@@ -2,15 +2,16 @@ Yet again it's been a while but here is new release of ddclient. As usual,
...
+       * added support for Yandex.Mail for Domain DNS service
+       * added support for NearlyFreeSpeech.net
+       * added support for DNS Made Easy
+       * added systemd instructions
+       * added support for dondominio.com
+       * updated perl instruction
+       * updated fritzbox instructions
+       * fixed multidomain support for namecheap
+       * fixed support for Yandex
+       * added support for OVH DynHost
```
Comment (SuperSandro2000) [inline RELEASENOTE:14]:
```suggestion
		* added support for Yandex.Mail for Domain DNS service
		* added support for NearlyFreeSpeech.net
		* added support for DNS Made Easy
		* added systemd instructions
		* added support for dondominio.com
		* updated perl instruction
		* updated fritzbox instructions
		* fixed multidomain support for namecheap
		* fixed support for Yandex
		* added support for OVH DynHost
```

File: RELEASENOTE [hunk 7]
```diff
@@ -2,15 +2,16 @@ Yet again it's been a while but here is new release of ddclient. As usual,
...
+		* added support for Yandex.Mail for Domain DNS service
+		* added support for NearlyFreeSpeech.net
+		* added support for DNS Made Easy
+		* added systemd instructions
+		* added support for dondominio.com
+		* updated perl instruction
+		* updated fritzbox instructions
+		* fixed multidomain support for namecheap
+		* fixed support for Yandex
+		* added support for OVH DynHost
```
Comment (SuperSandro2000) [inline RELEASENOTE:14]:
```suggestion
    * added support for Yandex.Mail for Domain DNS service
    * added support for NearlyFreeSpeech.net
    * added support for DNS Made Easy
    * added systemd instructions
    * added support for dondominio.com
    * updated perl instruction
    * updated fritzbox instructions
    * fixed multidomain support for namecheap
    * fixed support for Yandex
    * added support for OVH DynHost
```

**Comment by SuperSandro2000** [comment]
@ddclient/ddclientcore If no one has objections I will merge this.

I abused the suggestion & commit feature a bit to fix some whitespace changes and use a more standard password. I did not test this chance but as it won't break anything I think it is enough that the author tested it.

**Comment by xlammertink** [comment]
Hi @SuperSandro2000, thanks for approving. I have no objections, changes have indeed been tested.

File: ddclient [hunk 8]
```diff
@@ -5084,6 +5099,84 @@ sub nic_dnsmadeeasy_update {
     }
 }
 
+######################################################################
+## nic_ovh_examples
+######################################################################
+sub nic_ovh_examples {
+    return <<EoEXAMPLE;
```
Comment (SuperSandro2000) [inline ddclient:5106]:
```suggestion
    return <<'EoEXAMPLE';
```

File: RELEASENOTE [hunk 9]
```diff
@@ -2,15 +2,16 @@ Yet again it's been a while but here is new release of ddclient. As usual,
...
+    * added support for Yandex.Mail for Domain DNS service
+    * added support for NearlyFreeSpeech.net
+    * added support for DNS Made Easy
+    * added systemd instructions
+    * added support for dondominio.com
+    * updated perl instruction
+    * updated fritzbox instructions
+    * fixed multidomain support for namecheap
+    * fixed support for Yandex
+    * added support for OVH DynHost
```
Comment (SuperSandro2000) [inline RELEASENOTE:14]:
```suggestion
	* added support for Yandex.Mail for Domain DNS service
	* added support for NearlyFreeSpeech.net
	* added support for DNS Made Easy
	* added systemd instructions
	* added support for dondominio.com
	* updated perl instruction
	* updated fritzbox instructions
	* fixed multidomain support for namecheap
	* fixed support for Yandex
	* added support for OVH DynHost
```

**Comment by SuperSandro2000** [comment]
@rhansen Can you take a look at this, please?

I also plan to squash merge this to hide the not to nice commit history.

File: ddclient [hunk 10]
```diff
@@ -5084,6 +5099,84 @@ sub nic_dnsmadeeasy_update {
...
+
+        # Set the URL that we're going to update
+        my $url;
+        $url  = "https://";
+        $url .= "$config{$h}{'server'}$config{$h}{'script'}";
+        $url .= "?system=dyndns";
+        $url .= "&hostname=";
+        $url .= $h;
+        $url .= "&myip=";
+        $url .= $ip if $ip;
```
Comment (SuperSandro2000) [inline ddclient:5152]:
```suggestion
        $url .= "https://$config{$h}{'server'}$config{$h}{'script'}?system=dyndns";
        $url .= "&hostname=$h";
        $url .= "&myip=";
        $url .= $ip if $ip;
```

**Comment by rhansen** [comment]
> @rhansen Can you take a look at this, please?

I'll take a look soon—I wanna get through some of the other PRs first.

**Comment by xlammertink** [comment]
It is indeed very similar to other “protocols” as it’s a simple HTTPS request. I can recall minor differences in authentication (but it has been a while ago, so I’d have to double check). However, as there is not a clear HTTPS protocol I took the same approach as all other protocols that are already there. So in my opinion there should be either 1) a clear HTTPS protocol (which would require a rewrite of a lot of existing protocols) or 2) all domain registrars get a separated protocol.

**Comment by rhansen** [comment]
> It is indeed very similar to other “protocols” as it’s a simple HTTPS request. I can recall minor differences in authentication (but it has been a while ago, so I’d have to double check).

OK, then feel free to merge.

> there should be either 1) a clear HTTPS protocol (which would require a rewrite of a lot of existing protocols) or 2) all domain registrars get a separated protocol.

I think we can do something more incremental: Create a generic http(s) updater function and slowly migrate existing protocols to use that new function. That should eliminate a lot of the code duplication.

**Comment by SuperSandro2000** [comment]
How did you force push his branch?

**Comment by rhansen** [comment]
> @rhansen How did you force push his branch?

Just a normal `git push -f`. I was surprised it worked.

**Comment by SuperSandro2000** [comment]
> Just a normal `git push -f`. I was surprised it worked.

You pushed into his fork? Or did you check it out via gh and force pushed it there? I am very confused how you did this.

**Comment by rhansen** [comment]
> You pushed into his fork?

Yup, I pushed into the xlammertink fork. Here's how I did it:

```
git remote add xlammertink git@github.com:xlammertink/ddclient.git
git remote update -p
git checkout -b xlammertink-ovhsupport xlammertink/ovhsupport
git rebase -i origin/develop
git push -f
```

The "Allow edits by maintainers" checkbox says:

> If checked, users with write access to ddclient/ddclient can add new commits to your `foo` branch. You can always change this setting later.

Based on that wording I had assumed that GitHub would only allow fast-foward updates (add new commits), but I was delightfully surprised to see it accept force updates.

**Pull request merged**